docs/syntax.md

```diff
@@ -0,0 +1,255 @@
+
+```text
+module -> 'module' longmodulename { import } { statement }
+```
+
+The file must specify the module's `longmodulename` where its `modulename` must match the file's name not including the `.lbf` extension.
+After, the file may contain a sequence of `import`s followed by a sequence of `statement`s.
+
+### Import
+Imports bring *entities* (types and classes) of other modules into scope.
+
+```text
+import     -> 'import' [ 'qualified' ] longmodulename [ 'as' longmodulename ] [ importspec ]
+importspec -> '(' [ { typename ',' } typename [','] ] ')'
+```
+
+If `importspec` is omitted, then all entities specified in the module are imported; otherwise only the specified entities are imported.
+
+### Statement
+
+Statements define types, classes, instance clauses, and derive clauses.
+
+```text
+statement -> typedef
+           | classdef
+           | instanceclause
+           | deriveclause
+```
+
+#### Type definitions
+Types may be either sum types, product types, record types, or opaque types.
+
+```text
+typedef -> prodtypedef | sumtypedef |  recordtypedef | opaquetypedef
+```
+
+##### Product type definition
+A product type definition defines a new product type.
+
+```text
+prodtypedef -> 'prod' typename { varname } '=' prod
+prod        -> { typeexp }
+typeexp     -> varname
+             | longtypename
+             | '(' prod ')'
+```
+
+Product type definitions instruct the code generator to generate a product type for the target language.
+
+##### Sum type definition
+A sum type definition defines a new sum type.
+
+```text
+sumtypedef     -> 'sum' typename { varname } '=' sum
+sum            -> sumconstructor { '|' sumconstructor }
+sumconstructor -> typename prod
+```
+
+Sum type definitions instruct the code generator to generate a sum type for the target language.
+
+##### Record type definition
+A record type definition defines a new record type.
+
+```text
+recordtypedef -> 'record' typename { varname } '=' record
+record        -> '{' [ field { ',' field  } ] '}'
+field         -> fieldname ':' prod
+````
+
+Record type definitions instruct the code generator to generate a record type for the target language.
+
+##### Opaque type definition
+An opaque type definition defines a new opaque type.
+
+```text
+opaquetypedef -> 'opaque' typename { varname }
+```
+
+Opaque type definitions do not instruct the code generator to generate code, and an opaque type must be instead implemented in the target language.
+
+#### Class definition
+A class definition introduces a new class.
+
+```text
+classdef       -> 'class' [ constraintexps '<=' ] classname { varname }
+constraintexp  -> classref { varname }
+                | '(' constraintexps ')'
+constraintexps -> [ constraintexp { ',' constraintexp } ]
+```
+
+Class definitions do not instruct the code generator to generate code, but
+    instead provides a means to communicate with the code generator the
+    instances one would like to generate (via a derive clause).
```

Comment:
`communicate with the code generator the implementations that already exist (via instance clauses) or that one would like to generate (via derive clauses).`